### PR TITLE
Blog App: validations, model specs, and n+1 problems

### PR DESCRIPTION
**Pull request by hamid-murambiwa**
In this project I made the following changes:
- Added the following validations:
- [ ] For the User model:
  - Name must not be blank.
  - PostsCounter must be an integer greater than or equal to zero.
- [ ] For the Post model:
  - Title must not be blank.
  - Title must not exceed 250 characters.
  - CommentsCounter must be an integer greater than or equal to zero.
  - LikesCounter must be an integer greater than or equal to zero.
- Added unit specs for all of your models' methods and validations.
- Added flash messages in the create actions in all your controllers.
- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ def index
   end
 
   def show
-    @user = User.where(id: params[:id])
+    @user = User.find(params[:id])
     @posts = Post.where(id: params[:id])
```
Comment (rukundoeric) [inline app/controllers/users_controller.rb]: - [ ] <b>REQUIRED:</b> You can simply get the posts that belong to the user by : `@posts = @users.posts` and don't forget to fix N+1 problem which would be like this: `@posts = @users.posts.includes(:comments, :likes)`

**Review by rukundoeric — APPROVED** [review]
Hi @hamid-murambiwa, 

_Great job so far 👏 💯 👍. However, there are few issues we need to address before we move on. Please, check the comments below._
## Highlights
  - Follows git workflow.  💯
  - Codebase looks clean and professional. 💯 

## Required Changes ♻️
  - [ ] `posts#create`, `comments#create`, and `likes#create` actions uses flash messages for a successful and not successful path. That message is rendered in the view. Kindly do so to comply with project requirements.

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
  - [ ] Check the comments under the review.

Cheers and Happy coding!👏👏👏

__Feel free to leave any questions or comments in the PR thread if something is not 100% clear.__

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
> As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -5,7 +5,7 @@ def index
   end
 
   def show
-    @user = User.where(id: params[:id])
-    @posts = Post.where(id: params[:id])
+    @user = User.find(params[:id])
+    @posts = @user.posts.includes(:comments)
```
Comment (elisha2kyakpo1) [inline app/controllers/users_controller.rb]: [OPTIONAL]
- [ ] I suggest we take this piece of code in the index action. We will get user comments in our views very easily like so `@user.post.comments`

**Review by elisha2kyakpo1 — CHANGES_REQUESTED** [review]
Hi @hamid-murambiwa, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Required Changes ♻️
_Check the comments under the review._

- [ ] Let's address a change requested by the previous reviewer, a flash/notice from create actions show be rendered in the view(s) to let the user know if a post/comment or like was created or not.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Emmyn5600 — CHANGES_REQUESTED** [review]
Hi @hamid-murambiwa 

Good job so far! I know you have worked hard on making this milestone But  There are some issues that you still need to work on to go to the next project but you are almost there!

almost-there

## To highlight :champagne:
- All linters are passing :heavy_check_mark:
- Good commit history :heavy_check_mark:
- Good PR description :heavy_check_mark:

## Required Changes :recycle:

- [ ] The tests seem to be failing, kindly attach the screenshot for them as well as include the command you are using to run them.

image

- [ ] Kindly, Fix the below issue that is appearing in the attached image, you need to solve the N+1 problem when fetching all posts and their comments for a user as shown in the Example in the posts#index actions, you should use `@posts = @user.posts.includes(: comments)` instead of @posts = @user.posts.

image


## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!:clap::clap::clap:

#### Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request closed without merging**